pkg/cmd/cli/cmd/delete.go

```diff
@@ -0,0 +1,100 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	cmdutil "github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/resource"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+func NewCmdDelete(fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	longDesc := `Delete a resource by filename, stdin, resource and ID, or by resources and label selector.
+
+JSON and YAML formats are accepted.
+
+If both a filename and command line arguments are passed, the command line
+arguments are used and the filename is ignored.
+
+Note that the delete command does NOT do resource version checks, so if someone
+submits an update to a resource right when you submit a delete, their update
+will be lost along with the rest of the resource.
+
+Examples:
+
+	# Delete a pod using the type and ID specified in pod.json.
+	$ %[1]s delete -f pod.json
+
+	# Delete a pod based on the type and ID in the JSON passed into stdin.
+	$ cat pod.json | %[1]s delete -f -
+
+	# Delete pods and services with label name=myLabel.
+	$ %[1]s delete pods,services -l name=myLabel
+
+	# Delete a pod with ID 1234-56-7890-234234-456456.
+	$ %[1]s delete pod 1234-56-7890-234234-456456
+
+	# Delete all pods
+	$ %[1]s delete pods --all
+`
+	var filenames util.StringList
+	cmd := &cobra.Command{
+		Use:   "delete [-f FILENAME] | [RESOURCE (ID | -l label | --all)] | [-l label]",
+		Short: "Delete a resource by filename, stdin, resource and ID, or by resources and label selector.",
+		Long:  longDesc,
+		Run: func(cmd *cobra.Command, args []string) {
+			label := cmdutil.GetFlagString(cmd, "selector")
+			if len(args) == 0 && len(label) != 0 {
+				args = []string{"bc,builds,services,imageRepositories"}
```

Comment:
We should not hardcode this here.  The proper place to handle it is in expand resource shortcut code in factory.  Define one that lists all of the primary resources that an Openshift user would want and call it "all".  Don't hardcode it here.


> On Apr 2, 2015, at 8:55 AM, Andy Goldstein <notifications@github.com> wrote:
> 
> In pkg/cmd/cli/cmd/delete.go:
> 
> > +
> > +	# Delete a pod with ID 1234-56-7890-234234-456456.
> > +	$ %[1]s delete pod 1234-56-7890-234234-456456
> > +
> > +	# Delete all pods
> > +	$ %[1]s delete pods --all
> > +`
> > +	var filenames util.StringList
> > +	cmd := &cobra.Command{
> > +		Use:   "delete [-f FILENAME] | [RESOURCE (ID | -l label | --all)] | [-l label]",
> > +		Short: "Delete a resource by filename, stdin, resource and ID, or by resources and label selector.",
> > +		Long:  longDesc,
> > +		Run: func(cmd *cobra.Command, args []string) {
> > +			label := cmdutil.GetFlagString(cmd, "selector")
> > +			if len(args) == 0 && len(label) != 0 {
> > +				args = []string{"bc,builds,services,imageRepositories"}
> This is not the only place we have hardcoded lists of resources, and doing stuff like this makes things pretty brittle. As we add new resources or modify the system behavior, we'll have to update this and the other places (e.g. https://github.com/openshift/origin/blob/master/pkg/api/latest/latest.go#L70-L80, for just 1 example). Is there any way we can centralize this sort of code?
> 
> —
> Reply to this email directly or view it on GitHub.
>